From: Mike Snitzer <snitzer@redhat.com>
Date: Wed, 17 Feb 2010 00:31:40 -0500
Subject: [dm] raid1: fail writes if errors are not handled and log fails
Message-id: <1266366701-1575-5-git-send-email-snitzer@redhat.com>
Patchwork-id: 23306
O-Subject: [RHEL6 PATCH 4/5] dm raid1: fail writes if errors are not handled and
	log fails
Bugzilla: 565890
RH-Acked-by: Alasdair G Kergon <agk@redhat.com>

From: Mikulas Patocka <mpatocka@redhat.com>

Bugzilla #565890

commit 5528d17de1cf1462f285c40ccaf8e0d0e4c64dc0
Author: Mikulas Patocka <mpatocka@redhat.com>
Date:   Tue Feb 16 18:42:55 2010 +0000

    dm raid1: fail writes if errors are not handled and log fails

    If the mirror log fails when the handle_errors option was not selected
    and there is no remaining valid mirror leg, writes return success even
    though they weren't actually written to any device.  This patch
    completes them with EIO instead.

    This code path is taken:
    do_writes:
        bio_list_merge(&ms->failures, &sync);
    do_failures:
        if (!get_valid_mirror(ms)) (false)
        else if (errors_handled(ms)) (false)
        else bio_endio(bio, 0);

    The logic in do_failures is based on presuming that the write was already
    tried: if it succeeded at least on one leg (without handle_errors) it
    is reported as success.

    Reference: https://bugzilla.redhat.com/show_bug.cgi?id=555197

    Signed-off-by: Mikulas Patocka <mpatocka@redhat.com>
    Signed-off-by: Alasdair G Kergon <agk@redhat.com>

Signed-off-by: Aristeu Rozanski <arozansk@redhat.com>

diff --git a/drivers/md/dm-raid1.c b/drivers/md/dm-raid1.c
index 8aac410..33eba0a 100644
--- a/drivers/md/dm-raid1.c
+++ b/drivers/md/dm-raid1.c
@@ -731,7 +731,7 @@ static void do_writes(struct mirror_set *ms, struct bio_list *writes)
 	/*
 	 * Dispatch io.
 	 */
-	if (unlikely(ms->log_failure)) {
+	if (unlikely(ms->log_failure) && errors_handled(ms)) {
 		spin_lock_irq(&ms->lock);
 		bio_list_merge(&ms->failures, &sync);
 		spin_unlock_irq(&ms->lock);
